### PR TITLE
Support batch processing in UDF interface

### PR DESCRIPTION
**Pull request by Sh-Zh-7**
## Description
In this pr, we add a new batch processing UDF interface for scalar functions to avoid too much function calls and Java boxing/unboxing costs.
```java
void transform(Column[] columns, ColumnBuilder builder);
```
This modification has been observed to have huge performance benefit compared to old ones, which saves the end-to-end query time accordingly. More details can be seen in  https://lz8h91usu0.feishu.cn/docx/OA9tdC8wXow1b0xu6kOcPZFqnMg

Now IoTDB only calls the batch interface internally when dealing with scalar functions. For those User Defined Functions which only implement single row method, IoTDB can still handle them correctly by calling them row by row in new method's default implementation.

While this new method use tsfile module members like `Column` and `ColumnBuilder`, and introducing such a large dependency is unacceptable for users, for the packaging time will dramatically increases, we extract a `common-api` module to expose minimum required members from tsfile. And it will be shared by other modules in the future.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods.

### CODE REVIEW

File: iotdb-api/common-api/.gitignore [hunk 1]
```diff
@@ -0,0 +1,38 @@
+target/
+!.mvn/wrapper/maven-wrapper.jar
+!**/src/main/**/target/
+!**/src/test/**/target/
```
Comment (lancelly) [inline iotdb-api/common-api/.gitignore:4]: Delete this gitignore file？

File: iotdb-api/udf-api/pom.xml [hunk 2]
```diff
@@ -28,6 +28,13 @@
     </parent>
     <artifactId>udf-api</artifactId>
     <name>IoTDB: API: UDF API</name>
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.iotdb</groupId>
+            <artifactId>common-api</artifactId>
```
Comment (lancelly) [inline iotdb-api/udf-api/pom.xml:34]: udf-api

File: iotdb-api/common-api/pom.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <artifactId>iotdb-api</artifactId>
+        <groupId>org.apache.iotdb</groupId>
+        <version>1.3.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>common-api</artifactId>
+    <name>IoTDB: API: UDF API</name>
```
Comment (lancelly) [inline iotdb-api/common-api/pom.xml:10]: common api？

File: iotdb-api/udf-api/src/main/java/org/apache/iotdb/udf/api/UDTF.java [hunk 4]
```diff
@@ -112,6 +116,48 @@ default void transform(RowWindow rowWindow, PointCollector collector) throws Exc
...
+   */
+  default void transform(Column[] columns, ColumnBuilder builder) throws Exception {
+    int colCount = columns.length;
+    int rowCount = columns[0].getPositionCount();
+
+    // collect input data types from columns
+    TSDataType[] dataTypes = new TSDataType[colCount];
+    for (int i = 0; i < colCount; i++) {
+      dataTypes[i] = columns[i].getDataType();
+    }
```
Comment (lancelly) [inline iotdb-api/udf-api/src/main/java/org/apache/iotdb/udf/api/UDTF.java:139]: In my view, this interface, like other interfaces, should have a default empty implementation, allowing users to implement the specific interface. So I'm curious about the purpose of this implementation and RowImpl.

File: iotdb-api/common-api/.gitignore (hunk 1 above)
Comment (Sh-Zh-7) [reply iotdb-api/common-api/.gitignore]: Done.

File: iotdb-api/common-api/pom.xml (hunk 3 above)
Comment (Sh-Zh-7) [reply iotdb-api/common-api/pom.xml]: Done.

File: iotdb-api/udf-api/src/main/java/org/apache/iotdb/udf/api/UDTF.java (hunk 4 above)
Comment (lancelly) [reply iotdb-api/udf-api/src/main/java/org/apache/iotdb/udf/api/UDTF.java:139]: I see👍

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/SchemaQueryScanOperator.java [hunk 5]
```diff
@@ -41,11 +42,13 @@
...
 import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
-import static org.apache.iotdb.tsfile.read.common.block.TsBlockBuilderStatus.DEFAULT_MAX_TSBLOCK_SIZE_IN_BYTES;
 
 public class SchemaQueryScanOperator<T extends ISchemaInfo> implements SourceOperator {
+  private static final long MAX_SIZE =
+      TSFileDescriptor.getInstance().getConfig().getMaxTsBlockSizeInBytes();
 
-  private static final long MAX_SIZE = DEFAULT_MAX_TSBLOCK_SIZE_IN_BYTES;
+  private static final long DEFAULT_MAX_TSBLOCK_SIZE_IN_BYTES =
+      TSFileDescriptor.getInstance().getConfig().getMaxTsBlockSizeInBytes();
```
Comment (lancelly) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/SchemaQueryScanOperator.java:51]: redundant variable?

Comment (Sh-Zh-7) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/SchemaQueryScanOperator.java:51]: Both MAX_SIZE and DEFAULT_MAX_TSBLOCK_SIZE_IN_BYTES are used.

Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/SchemaQueryScanOperator.java:51]: Only one of them is necessary.

Comment (Sh-Zh-7) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/schema/SchemaQueryScanOperator.java]: Done.

**Review by lancelly — APPROVED** [review]
LGTM

**Review by Sh-Zh-7 — COMMENTED** [review]
Double check code finished.

**Pull request merged**